concantenate LICENSE files when building a wheel

mattip opened the pull request:
Fixes #50695

I checked locally that the concatenated license file appears at `torch-<version>.dist-info/LICENSE` in the wheel.

mattip committed:
concantenate LICENSE files when building a wheel

mattip commented:
@rgommers thoughts?

walterddr reviewed (commented):
looks good. would it be possible to add a test in `test/test_license.py` ?

rgommers commented on setup.py:
```diff
@@ -552,6 +552,48 @@ def load(filename):
+    """Merge LICENSE.txt and LICENSES_BUNDLED.txt as a context manager
+
+    Done this way to keep LICENSE.txt in repo as exact BSD 3-clause.
+    This makes GitHub web gui state correctly how the repo is licensed.
```
These two sentences aren't true, I'd fix that. The main `LICENSE.txt` file needs to be exactly BSD-3 for it to be recognized, which I don't think you can do - there's too much copyright info in it for example.

This should say something like:
_`LICENSE.txt` is the main PyTorch license, `LICENSES_BUNDLED.txt` is auto-generated from all the licenses found in third_party/._

rgommers commented on setup.py:
```diff
@@ -552,6 +552,48 @@ def load(filename):
+try:
+    from wheel.bdist_wheel import bdist_wheel
+except ImportError:
+    # This is useful when wheel is not isntalled and bdist_wheel is not
```
typo in isntalled

rgommers reviewed (changes requested):
LGTM modulo the docs issues, thanks @mattip.

mattip committed:
fixes from review, add test

mattip commented:
Added a test that iff torch is installed (there is a `torch-*dist-info` directory in `site-packages`) the bundled licenses appear in the LICENSE file. Also fixed comments from review.

mattip committed:
fix fstring

rgommers commented on test/test_license.py:
```diff
@@ -26,5 +30,20 @@ def test_license_in_wheel(self):
+        site-package/torch-*dist-info/LICENSE. Make sure it contains the third
+        party bundle of licenses"""
+
+        site_packages = os.path.dirname(os.path.dirname(torch.__file__))
```
This dir would be `test/`, and tests aren't installed to begin with, so this cannot work?

mattip replied:
The test is intended for the CI runs which used the wheel to install. Do they install to `test/` rather than `site-packages` when doing `pip/conda install` ?

rgommers replied:
If "they" is tests - they don't get installed at all. Note the directory layout:
```
torch/
test/
``
not `torch/test/`. There is nowhere for the tests to go. In CI, tests live in the cloned repo, while `torch/` can get installed to `site-packages`.

rgommers replied:
Oh wait, I may be confused. Let me re-read the code.

rgommers replied:
Ah okay, sorry my comment was confusing.

This test should indeed work, but not always. There's multiple ways to install, and `site_packages/torch-*dist-info/` is not guaranteed to exist, so I think you should skip the test if it doesn't.

walterddr replied:
One more thing: In order for CI to pick up your test file you need to add it to the list of `TESTS = [ ... ` in test/run_test.py
currently it is not running on CI i am afraid.

samestep replied:
@walterddr Should we add a CI check that warns on files in `test` that aren't actually being run? Perhaps add another list in `test/run_test.py` that includes all the files that aren't tested, to ensure that any new file in `test` gets added to one of those lists?

mattip replied:
refactored the test to skip instead of silently pass, and added the file to `test/run_test.py`

walterddr replied:
@samestep  good point we can follow up

mattip committed:
fixes from review

seemethere commented on test/run_test.py:
```diff
@@ -108,6 +108,11 @@
     'test_functional_autograd_benchmark',
     'test_package',
+    'test_license',
+:qa
```
I'm assuming this isn't meant to be there?

mattip replied:
whoops, fixed

seemethere reviewed (commented):
Looks mostly good to me just one thing left to clean up!

mattip committed:
typo

mattip commented:
CI failure seems unrelated. I think this is ready for final review.
